// Code generated by protoc-gen-validate. DO NOT EDIT.
// source: party_affairs_resource_classify.proto

package v1

import (
	"bytes"
	"errors"
	"fmt"
	"net"
	"net/mail"
	"net/url"
	"regexp"
	"sort"
	"strings"
	"time"
	"unicode/utf8"

	"google.golang.org/protobuf/types/known/anypb"
)

// ensure the imports are used
var (
	_ = bytes.MinRead
	_ = errors.New("")
	_ = fmt.Print
	_ = utf8.UTFMax
	_ = (*regexp.Regexp)(nil)
	_ = (*strings.Reader)(nil)
	_ = net.IPv4len
	_ = time.Duration(0)
	_ = (*url.URL)(nil)
	_ = (*mail.Address)(nil)
	_ = anypb.Any{}
	_ = sort.Sort
)

// Validate checks the field values on ResourceClassify with the rules defined
// in the proto definition for this message. If any rules are violated, the
// first error encountered is returned, or nil if there are no violations.
func (m *ResourceClassify) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ResourceClassify with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// ResourceClassifyMultiError, or nil if none found.
func (m *ResourceClassify) ValidateAll() error {
	return m.validate(true)
}

func (m *ResourceClassify) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Id

	// no validation rules for Name

	// no validation rules for CreatedAt

	// no validation rules for UpdatedAt

	if m.Weight != nil {
		// no validation rules for Weight
	}

	if len(errors) > 0 {
		return ResourceClassifyMultiError(errors)
	}

	return nil
}

// ResourceClassifyMultiError is an error wrapping multiple validation errors
// returned by ResourceClassify.ValidateAll() if the designated constraints
// aren't met.
type ResourceClassifyMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ResourceClassifyMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ResourceClassifyMultiError) AllErrors() []error { return m }

// ResourceClassifyValidationError is the validation error returned by
// ResourceClassify.Validate if the designated constraints aren't met.
type ResourceClassifyValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ResourceClassifyValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ResourceClassifyValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ResourceClassifyValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ResourceClassifyValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ResourceClassifyValidationError) ErrorName() string { return "ResourceClassifyValidationError" }

// Error satisfies the builtin error interface
func (e ResourceClassifyValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sResourceClassify.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ResourceClassifyValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ResourceClassifyValidationError{}

// Validate checks the field values on AllResourceClassifyReply with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *AllResourceClassifyReply) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on AllResourceClassifyReply with the
// rules defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// AllResourceClassifyReplyMultiError, or nil if none found.
func (m *AllResourceClassifyReply) ValidateAll() error {
	return m.validate(true)
}

func (m *AllResourceClassifyReply) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	for idx, item := range m.GetList() {
		_, _ = idx, item

		if all {
			switch v := interface{}(item).(type) {
			case interface{ ValidateAll() error }:
				if err := v.ValidateAll(); err != nil {
					errors = append(errors, AllResourceClassifyReplyValidationError{
						field:  fmt.Sprintf("List[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			case interface{ Validate() error }:
				if err := v.Validate(); err != nil {
					errors = append(errors, AllResourceClassifyReplyValidationError{
						field:  fmt.Sprintf("List[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			}
		} else if v, ok := interface{}(item).(interface{ Validate() error }); ok {
			if err := v.Validate(); err != nil {
				return AllResourceClassifyReplyValidationError{
					field:  fmt.Sprintf("List[%v]", idx),
					reason: "embedded message failed validation",
					cause:  err,
				}
			}
		}

	}

	if len(errors) > 0 {
		return AllResourceClassifyReplyMultiError(errors)
	}

	return nil
}

// AllResourceClassifyReplyMultiError is an error wrapping multiple validation
// errors returned by AllResourceClassifyReply.ValidateAll() if the designated
// constraints aren't met.
type AllResourceClassifyReplyMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m AllResourceClassifyReplyMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m AllResourceClassifyReplyMultiError) AllErrors() []error { return m }

// AllResourceClassifyReplyValidationError is the validation error returned by
// AllResourceClassifyReply.Validate if the designated constraints aren't met.
type AllResourceClassifyReplyValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e AllResourceClassifyReplyValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e AllResourceClassifyReplyValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e AllResourceClassifyReplyValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e AllResourceClassifyReplyValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e AllResourceClassifyReplyValidationError) ErrorName() string {
	return "AllResourceClassifyReplyValidationError"
}

// Error satisfies the builtin error interface
func (e AllResourceClassifyReplyValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sAllResourceClassifyReply.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = AllResourceClassifyReplyValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = AllResourceClassifyReplyValidationError{}

// Validate checks the field values on AddResourceClassifyRequest with the
// rules defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *AddResourceClassifyRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on AddResourceClassifyRequest with the
// rules defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// AddResourceClassifyRequestMultiError, or nil if none found.
func (m *AddResourceClassifyRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *AddResourceClassifyRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if utf8.RuneCountInString(m.GetName()) < 1 {
		err := AddResourceClassifyRequestValidationError{
			field:  "Name",
			reason: "value length must be at least 1 runes",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	if m.GetWeight() < 0 {
		err := AddResourceClassifyRequestValidationError{
			field:  "Weight",
			reason: "value must be greater than or equal to 0",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	if len(errors) > 0 {
		return AddResourceClassifyRequestMultiError(errors)
	}

	return nil
}

// AddResourceClassifyRequestMultiError is an error wrapping multiple
// validation errors returned by AddResourceClassifyRequest.ValidateAll() if
// the designated constraints aren't met.
type AddResourceClassifyRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m AddResourceClassifyRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m AddResourceClassifyRequestMultiError) AllErrors() []error { return m }

// AddResourceClassifyRequestValidationError is the validation error returned
// by AddResourceClassifyRequest.Validate if the designated constraints aren't met.
type AddResourceClassifyRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e AddResourceClassifyRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e AddResourceClassifyRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e AddResourceClassifyRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e AddResourceClassifyRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e AddResourceClassifyRequestValidationError) ErrorName() string {
	return "AddResourceClassifyRequestValidationError"
}

// Error satisfies the builtin error interface
func (e AddResourceClassifyRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sAddResourceClassifyRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = AddResourceClassifyRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = AddResourceClassifyRequestValidationError{}

// Validate checks the field values on UpdateResourceClassifyRequest with the
// rules defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *UpdateResourceClassifyRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on UpdateResourceClassifyRequest with
// the rules defined in the proto definition for this message. If any rules
// are violated, the result is a list of violation errors wrapped in
// UpdateResourceClassifyRequestMultiError, or nil if none found.
func (m *UpdateResourceClassifyRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *UpdateResourceClassifyRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if m.GetId() <= 0 {
		err := UpdateResourceClassifyRequestValidationError{
			field:  "Id",
			reason: "value must be greater than 0",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	if utf8.RuneCountInString(m.GetName()) < 1 {
		err := UpdateResourceClassifyRequestValidationError{
			field:  "Name",
			reason: "value length must be at least 1 runes",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	if m.GetWeight() < 0 {
		err := UpdateResourceClassifyRequestValidationError{
			field:  "Weight",
			reason: "value must be greater than or equal to 0",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	if len(errors) > 0 {
		return UpdateResourceClassifyRequestMultiError(errors)
	}

	return nil
}

// UpdateResourceClassifyRequestMultiError is an error wrapping multiple
// validation errors returned by UpdateResourceClassifyRequest.ValidateAll()
// if the designated constraints aren't met.
type UpdateResourceClassifyRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m UpdateResourceClassifyRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m UpdateResourceClassifyRequestMultiError) AllErrors() []error { return m }

// UpdateResourceClassifyRequestValidationError is the validation error
// returned by UpdateResourceClassifyRequest.Validate if the designated
// constraints aren't met.
type UpdateResourceClassifyRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e UpdateResourceClassifyRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e UpdateResourceClassifyRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e UpdateResourceClassifyRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e UpdateResourceClassifyRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e UpdateResourceClassifyRequestValidationError) ErrorName() string {
	return "UpdateResourceClassifyRequestValidationError"
}

// Error satisfies the builtin error interface
func (e UpdateResourceClassifyRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sUpdateResourceClassifyRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = UpdateResourceClassifyRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = UpdateResourceClassifyRequestValidationError{}

// Validate checks the field values on DeleteResourceClassifyRequest with the
// rules defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *DeleteResourceClassifyRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on DeleteResourceClassifyRequest with
// the rules defined in the proto definition for this message. If any rules
// are violated, the result is a list of violation errors wrapped in
// DeleteResourceClassifyRequestMultiError, or nil if none found.
func (m *DeleteResourceClassifyRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *DeleteResourceClassifyRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if m.GetId() <= 0 {
		err := DeleteResourceClassifyRequestValidationError{
			field:  "Id",
			reason: "value must be greater than 0",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	if len(errors) > 0 {
		return DeleteResourceClassifyRequestMultiError(errors)
	}

	return nil
}

// DeleteResourceClassifyRequestMultiError is an error wrapping multiple
// validation errors returned by DeleteResourceClassifyRequest.ValidateAll()
// if the designated constraints aren't met.
type DeleteResourceClassifyRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m DeleteResourceClassifyRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m DeleteResourceClassifyRequestMultiError) AllErrors() []error { return m }

// DeleteResourceClassifyRequestValidationError is the validation error
// returned by DeleteResourceClassifyRequest.Validate if the designated
// constraints aren't met.
type DeleteResourceClassifyRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e DeleteResourceClassifyRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e DeleteResourceClassifyRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e DeleteResourceClassifyRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e DeleteResourceClassifyRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e DeleteResourceClassifyRequestValidationError) ErrorName() string {
	return "DeleteResourceClassifyRequestValidationError"
}

// Error satisfies the builtin error interface
func (e DeleteResourceClassifyRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sDeleteResourceClassifyRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = DeleteResourceClassifyRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = DeleteResourceClassifyRequestValidationError{}
